File: go/pkg/vecstore/memory.go
package vecstore

import (
	"fmt"
	"math"
	"sort"
	"sync"
)

// Memory is an in-memory Index implementation using brute-force cosine
// distance. Intended for testing and small-scale use (< 1000 vectors).
//
// It is safe for concurrent use.
type Memory struct {
	mu      sync.RWMutex
	vectors map[string][]float32
}

// NewMemory creates a new in-memory vector index.
func NewMemory() *Memory {
	return &Memory{
		vectors: make(map[string][]float32),
	}
}

func (m *Memory) Insert(id string, vector []float32) error {
	cp := make([]float32, len(vector))
	copy(cp, vector)
	m.mu.Lock()
	m.vectors[id] = cp
	m.mu.Unlock()
	return nil
}

func (m *Memory) BatchInsert(ids []string, vectors [][]float32) error {
	if len(ids) != len(vectors) {
		return fmt.Errorf("vecstore: BatchInsert length mismatch: %d ids, %d vectors", len(ids), len(vectors))
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	for i, id := range ids {
		cp := make([]float32, len(vectors[i]))
		copy(cp, vectors[i])
		m.vectors[id] = cp
	}
	return nil
}

func (m *Memory) Flush() error {
	return nil // in-memory: always visible
}

func (m *Memory) Search(query []float32, topK int) ([]Match, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if len(m.vectors) == 0 || topK <= 0 {
		return nil, nil
	}

	type scored struct {
		id   string
		dist float32
	}
	results := make([]scored, 0, len(m.vectors))
	for id, vec := range m.vectors {
		d := CosineDistance(query, vec)
		results = append(results, scored{id: id, dist: d})
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].dist < results[j].dist
	})

	if len(results) > topK {
		results = results[:topK]
	}

	matches := make([]Match, len(results))
	for i, r := range results {
		matches[i] = Match{ID: r.id, Distance: r.dist}
	}
	return matches, nil
}

func (m *Memory) Delete(id string) error {
	m.mu.Lock()
	delete(m.vectors, id)
	m.mu.Unlock()
	return nil
}

func (m *Memory) Len() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.vectors)
}

func (m *Memory) Close() error {
	return nil
}

// CosineDistance computes the cosine distance between two vectors.
// Returns a value in [0, 2] where 0 means identical direction and
// 2 means opposite direction. Returns 0 if either vector has zero norm.
func CosineDistance(a, b []float32) float32 {
	if len(a) != len(b) {
		return 2 // maximum distance for mismatched dimensions
	}

	var dot, normA, normB float64
	for i := range a {
		ai, bi := float64(a[i]), float64(b[i])
		dot += ai * bi
		normA += ai * ai
		normB += bi * bi
	}

	if normA == 0 || normB == 0 {
		return 2 // zero vector has no direction; treat as maximum distance
	}

	similarity := dot / (math.Sqrt(normA) * math.Sqrt(normB))
	// Clamp to [-1, 1] to handle floating point errors.
	if similarity > 1 {
		similarity = 1
	}
	if similarity < -1 {
		similarity = -1
	}
	return float32(1 - similarity)
}

File: rust/vecstore/src/memory.rs
use std::collections::HashMap;
use std::sync::RwLock;

use crate::cosine::cosine_distance;
use crate::error::VecError;
use crate::vecstore::{Match, VecIndex};

/// Memory is an in-memory VecIndex using brute-force cosine distance.
/// Intended for testing and small-scale use (< 1000 vectors).
pub struct MemoryIndex {
    vectors: RwLock<HashMap<String, Vec<f32>>>,
}

impl MemoryIndex {
    pub fn new() -> Self {
        Self {
            vectors: RwLock::new(HashMap::new()),
        }
    }
}

impl Default for MemoryIndex {
    fn default() -> Self {
        Self::new()
    }
}

impl VecIndex for MemoryIndex {
    fn insert(&self, id: &str, vector: &[f32]) -> Result<(), VecError> {
        let mut vecs = self.vectors.write().unwrap();
        vecs.insert(id.to_string(), vector.to_vec());
        Ok(())
    }

    fn batch_insert(&self, ids: &[&str], vectors: &[&[f32]]) -> Result<(), VecError> {
        if ids.len() != vectors.len() {
            return Err(VecError::BatchLengthMismatch {
                ids: ids.len(),
                vectors: vectors.len(),
            });
        }
        let mut vecs = self.vectors.write().unwrap();
        for (id, vec) in ids.iter().zip(vectors.iter()) {
            vecs.insert(id.to_string(), vec.to_vec());
        }
        Ok(())
    }

    fn search(&self, query: &[f32], top_k: usize) -> Result<Vec<Match>, VecError> {
        let vecs = self.vectors.read().unwrap();
        if vecs.is_empty() || top_k == 0 {
            return Ok(vec![]);
        }

        let mut results: Vec<(String, f32)> = vecs
            .iter()
            .map(|(id, vec)| (id.clone(), cosine_distance(query, vec)))
            .collect();

        results.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));

        if results.len() > top_k {
            results.truncate(top_k);
        }

        Ok(results
            .into_iter()
            .map(|(id, distance)| Match { id, distance })
            .collect())
    }

    fn delete(&self, id: &str) -> Result<(), VecError> {
        let mut vecs = self.vectors.write().unwrap();
        vecs.remove(id);
        Ok(())
    }

    fn len(&self) -> usize {
        self.vectors.read().unwrap().len()
    }

    fn flush(&self) -> Result<(), VecError> {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_search() {
        let idx = MemoryIndex::new();
        idx.insert("a", &[1.0, 0.0, 0.0, 0.0]).unwrap();
        idx.insert("b", &[0.0, 1.0, 0.0, 0.0]).unwrap();
        idx.insert("c", &[0.9, 0.1, 0.0, 0.0]).unwrap();

        let matches = idx.search(&[1.0, 0.0, 0.0, 0.0], 2).unwrap();
        assert_eq!(matches.len(), 2);
        assert_eq!(matches[0].id, "a");
        assert_eq!(matches[1].id, "c");
    }

    #[test]
    fn test_batch_insert() {
        let idx = MemoryIndex::new();
        idx.batch_insert(
            &["a", "b", "c"],
            &[&[1.0, 0.0, 0.0], &[0.0, 1.0, 0.0], &[0.0, 0.0, 1.0]],
        )
        .unwrap();
        assert_eq!(idx.len(), 3);

        let matches = idx.search(&[1.0, 0.0, 0.0], 1).unwrap();
        assert_eq!(matches.len(), 1);
        assert_eq!(matches[0].id, "a");
    }

    #[test]
    fn test_batch_insert_mismatch() {
        let idx = MemoryIndex::new();
        assert!(idx
            .batch_insert(&["a", "b"], &[&[1.0, 0.0]])
            .is_err());
    }

    #[test]
    fn test_delete() {
        let idx = MemoryIndex::new();
        idx.insert("a", &[1.0, 0.0]).unwrap();
        assert_eq!(idx.len(), 1);
        idx.delete("a").unwrap();
        assert_eq!(idx.len(), 0);
        idx.delete("nonexistent").unwrap();
    }

    #[test]
    fn test_search_empty() {
        let idx = MemoryIndex::new();
        let matches = idx.search(&[1.0, 0.0, 0.0], 5).unwrap();
        assert!(matches.is_empty());
    }

    #[test]
    fn test_flush() {
        assert!(MemoryIndex::new().flush().is_ok());
    }
}
